src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala

```diff
@@ -1860,3 +1880,184 @@ object StorageUtils {
+      meta.setContentLength(dataSizeInBuffer)
+      byteArrayOutputStream.close()
+      val buffer = byteArrayOutputStream.toByteArray
+      val inputStream: InputStream =
+        new ByteArrayInputStream(buffer, 0, dataSizeInBuffer)
+      s3Client.putObject(bucketName, keyName, inputStream, meta)
+      val usedTime = System.currentTimeMillis() - start
+      totalUploadTime += usedTime
+
+      CloudStorageOperations.log.info(
+        s"""${SnowflakeResultSetRDD.WORKER_LOG_PREFIX}:
+           | Finish uploading file $file without AWS multiple parts API
+           | because the data size is less than the buffer size:
+           | bufferSize=${Utils.getSizeString(bufferSize)}
+           | compressedSize=${Utils.getSizeString(dataSizeInBuffer)}
+           |""".stripMargin.filter(_ >= ' '))
+    } else {
+      // Upload the data chunk with Multiple block upload API
+      doUploadOnePart()
+
+      // For last block, commit the multipart upload.
+      if (isLastChunk) {
+        byteArrayOutputStream.close()
+        val start = System.currentTimeMillis()
+        val compRequest = new CompleteMultipartUploadRequest(
+          bucketName, keyName, initResponse.getUploadId, partETags)
+        s3Client.completeMultipartUpload(compRequest)
+        val usedTime = System.currentTimeMillis() - start
+        totalUploadTime += usedTime
+        CloudStorageOperations.log.info(
+          s"""${SnowflakeResultSetRDD.WORKER_LOG_PREFIX}:
+             | Finish uploading file $file with AWS multiple parts API:
+             | bufferSize=${Utils.getSizeString(bufferSize)} chunkCount=$partId
+             | compressedSize=${Utils.getSizeString(totalDataSize)}
+             |""".stripMargin.filter(_ >= ' '))
+      }
+    }
+  }
+
+  // upload one part with multiple part upload API
+  private def doUploadOnePart(): Unit = {
+    val start = System.currentTimeMillis()
+    // Indicate the upload to using multiple parts API.
+    useMultipleBlockUpload = true
+    // byteArrayOutputStream.close()
+    val buffer = byteArrayOutputStream.toByteArray
+    val inputStream: InputStream =
+      new ByteArrayInputStream(buffer, 0, dataSizeInBuffer)
+
+    // Create the request to upload a part.
+    val uploadRequest = new UploadPartRequest()
+      .withBucketName(bucketName)
+      .withKey(keyName)
+      .withUploadId(initResponse.getUploadId)
+      .withPartNumber(partId)
+      .withInputStream(inputStream)
+      .withPartSize(dataSizeInBuffer)
+
+    // Upload the part and add the response's ETag to our list.
+    val uploadResult = s3Client.uploadPart(uploadRequest)
+    partETags.add(uploadResult.getPartETag)
+
+    // Update status
+    totalDataSize += dataSizeInBuffer
+    dataSizeInBuffer = 0
+    partId += 1
+    byteArrayOutputStream.reset()
+    // byteArrayOutputStream = new ByteArrayOutputStream(bufferSize + 8)
+
+    val usedTime = System.currentTimeMillis() - start
+    totalUploadTime += usedTime
+
+    if (partId == 3) {
```

Comment:
?

Reply:
I guess your question is why "if (partId == 3)" is used. correct?
It only injects an exception when uploading the 2nd part.

Reply:
There should be no difference to inject an exception with other parts. I thought it may be more common to hit upload error on non-first part.

Reply:
I guess my confusion is why this piece of code that's for testing on the same code path as production code?

Reply:
Thanks for the clarification.
This is used to simulate an upload error by raising an exception if the flag is ON.
The flag is enabled only if the some test. For all the others, there is an extra IF check. 
But, the perf side-effects for production code can be ignored.

Reply:
Ah I see, I guess it just still seems a bit weird to have this on the same code path as regular production. Is there any other way to raise an exception reliably for your test?

Reply:
In C/C++, we can use conditional compilation such as "#ifdef ... #endif".
But, I didn't find the equivalent functionality in Java/Scala. If you or bing have any idea about this, please let me know. I can refactor it.

Reply:
How about this:

Close the stream from your test after uploading two parts. If your stream is closed, trying to upload more parts should raise an exception.

Reply:
The test case is writing a data frame to a snowflake table. It is difficult because
1. It needs some hacking to let the test to get the stream object.
2. The test is difficult to catch the timing to close the stream.